### PR TITLE
Nvfuser - Type Promotion Fix

### PR DESCRIPTION
**Pull request by rdspring1**
Fix Type Promotion failures in [Issue 76046](https://github.com/pytorch/pytorch/issues/76046)

1. Updated nvfuser type promotion rule for codegen kernel;
2. Updated casting for output of nvfuser kernel to respect profiling/TorchScript scalar type;
3. Updated type_inference.cpp to only update device/scalar_type when profiling information is missing.

Additional Type Promotion Fixes:
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float32
-  test_nvfuser_correctness_softmax_with_dtype_cuda_bfloat16
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float16
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float32
-  test_nvfuser_correctness_log_softmax_dtype_cuda_bfloat16
-  test_nvfuser_correctness_log_softmax_dtype_cuda_bool
-  test_nvfuser_correctness_log_softmax_dtype_cuda_float16
-  test_nvfuser_correctness_log_softmax_dtype_cuda_float32
-  test_nvfuser_correctness_sum_cuda_int32
-  test_nvfuser_correctness_sum_to_size_cuda_int32

### CODE REVIEW

**Comment by mruberry** [comment]
Lintrunner failure is real. We have a utility to help run it locally now: https://github.com/pytorch/pytorch/wiki/lintrunner

**Comment by mruberry** [comment]
@jjsjann123 -- ping me when you're happy with this?

File: torch/csrc/jit/codegen/cuda/arith.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1009]: DITTO

File: torch/csrc/jit/codegen/cuda/arith.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1060]: DITTO

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1970]: Note: updated math type... we could stick with our math type in parser.

File: torch/csrc/jit/codegen/cuda/arith.cpp
Comment (mruberry) [inline torch/csrc/jit/codegen/cuda/arith.cpp:189]: Not a requirement, but a brief comment for each function would be nice for readability/reviewability

File: torch/csrc/jit/codegen/cuda/arith.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp:764]: We need to have better support for dtype argument in integration. e.g. currently some parse entry (i.e. sum) only fuses node when dtype are unspecified. :stuck_out_tongue:

**Review by jjsjann123 — APPROVED** [review]
LGTM.

**Review by mruberry — APPROVED** [review]
Stamped! Nice OpInfo skip fixes.

**Comment by mruberry** [comment]
@rdspring1 @jjsjann123 I think you can merge these PRs yourselves now using @pytochbot merge this please when you're ready. It just needs an approving review from someone with commit writes

**Pull request closed without merging**
